src/core/hle/kernel/memory/page_table.cpp

```diff
@@ -0,0 +1,1230 @@
+
+    if (const ResultCode result{CheckMemoryState(
+            &prev_state, &prev_perm, nullptr, addr, size, MemoryState::FlagCode,
+            MemoryState::FlagCode, MemoryPermission::None, MemoryPermission::None,
+            MemoryAttribute::Mask, MemoryAttribute::None, MemoryAttribute::IpcAndDeviceMapped)};
+        result.IsError()) {
+        return result;
+    }
+
+    MemoryState state{prev_state};
+
+    // Ensure state is mutable if permission allows write
+    if ((perm & MemoryPermission::Write) != MemoryPermission::None) {
+        if (prev_state == MemoryState::Code) {
+            state = MemoryState::CodeData;
+        } else if (prev_state == MemoryState::AliasCode) {
+            state = MemoryState::AliasCodeData;
+        } else {
+            UNREACHABLE();
+        }
+    }
+
+    // Return early if there is nothing to change
+    if (state == prev_state && perm == prev_perm) {
+        return RESULT_SUCCESS;
+    }
+
+    const std::size_t num_pages{size / PageSize};
+    const OperationType operation{(perm & MemoryPermission::Execute) != MemoryPermission::None
+                                      ? OperationType::ChangePermissionsAndRefresh
+                                      : OperationType::ChangePermissions};
+
+    if (const ResultCode result{Operate(addr, num_pages, perm, operation)}; result.IsError()) {
+        return result;
+    }
+
+    block_manager->Update(addr, num_pages, state, perm);
+
+    return RESULT_SUCCESS;
+}
+
+MemoryInfo PageTable::QueryInfoImpl(VAddr addr) {
+    std::lock_guard<std::recursive_mutex> lock{page_table_lock};
+
+    return block_manager->FindBlock(addr).GetMemoryInfo();
+}
+
+MemoryInfo PageTable::QueryInfo(VAddr addr) {
+    if (!Contains(addr, 1)) {
+        return {address_space_end,      0 - address_space_end, MemoryState::Inaccessible,
+                MemoryPermission::None, MemoryAttribute::None, MemoryPermission::None};
+    }
+
+    return QueryInfoImpl(addr);
+}
+
+ResultCode PageTable::ReserveTransferMemory(VAddr addr, std::size_t size, MemoryPermission perm) {
+    std::lock_guard<std::recursive_mutex> lock{page_table_lock};
+
+    MemoryState state{};
+    MemoryAttribute attribute{};
+
+    if (const ResultCode result{CheckMemoryState(
+            &state, nullptr, &attribute, addr, size,
+            MemoryState::FlagCanTransfer | MemoryState::FlagReferenceCounted,
+            MemoryState::FlagCanTransfer | MemoryState::FlagReferenceCounted,
+            MemoryPermission::Mask, MemoryPermission::ReadAndWrite, MemoryAttribute::Mask,
+            MemoryAttribute::None, MemoryAttribute::IpcAndDeviceMapped)};
+        result.IsError()) {
+        return result;
+    }
+
+    block_manager->Update(addr, size / PageSize, state, perm, attribute | MemoryAttribute::Locked);
+
+    return RESULT_SUCCESS;
+}
+
+ResultCode PageTable::ResetTransferMemory(VAddr addr, std::size_t size) {
+    std::lock_guard<std::recursive_mutex> lock{page_table_lock};
```

Comment:
```suggestion
    std::lock_guard lock{page_table_lock};
```

etc for other occurences in this file.